### ISSUE TITLE
MEP26 Artist Stylesheet - Tracking Issue

### ISSUE THREAD

**Issue by OceanWolf** [matplotlib/matplotlib#6550]
Since the move of putting MEPs into the matplotlib documentation through PRs instead of using the wiki, I think we also need issues for discussion across PRs, so here goes one for MEP26.

Attn: @JamesRamm

One of the points in the methodology says "A mechanism by which to parse a stylesheet into a tree", last night I began looking into libraries for this and came across https://github.com/SimonSapin/tinycss2 this appears to do that for us, but I still have to play with it.

Personally I think we will need to generalise it a bit more so we can disallow some of selector rules which obviously don't apply to MPL AFAIK; and adding some kind of difference, to modify rules from code.

What do other people think?

**Comment by reedstrm** [comment]
@OceanWolf saw your comments in the issue on Simon's cssselect2 repo. I've been building some actual code to use an (extended) CSS syntax to do for HTML basically what XSLT does for XML (actually move content around), over at Connexions/cnx-easybake. I've offered to take over cssselect2 from Simon, and am right now working on a way to expose extension hooks - parser, compiler, etc. The point is not only to save rewriting the world, but to allow our non-programmer content people to leverage knowledge of CSS to design other, structural aspects of our documents. In our specific case, text books, designing things like auto-generated table of content, with numbering spanning multiple HTML docs, or moving problems into a problem set at the end of a chapter, but move the answers to the end of a book. 

I did a proof-of-concept w/ cssselect, but the internal design (mapping CSS selectors to xpaths) not only has bad corner cases where the models don't match, but also suffers from performance issues. Re-implementing in cssselect2 brought the test case down from 20-30min to < 2min. (numbering and rearranging and entire Physics textbook).

I have a slightly hacked up version of cssselect2 (Connexions/cssselect2), but want to instead incorporate an extension mechanism, so the library can run in a strict standards compliant mode, or be given a dictionary of extensions to recognize in addition. I'd love to have other use-cases than mine, to see if I can make a truly generalized, but usable, extension mechanism.

**Comment by OceanWolf** [comment]
Good to hear from you @reedstrm, I don't quite understand what you mean by creating an XCSS for HTML as XSLT for XML, do you mean you write a non-css that styles HTML?  Do you mean something like [SASS](http://sass-lang.com/)?  I would love to see it parse a SASS file directly rather then having to compile it first.  Would this get done in ``cssselect2``, or ``tinycss2``?

I have just got it working (with some hacking so I don't have to install ``webencodings``.  I wonder if we can remove that dependency (one extra package to get into the repos, as well as adding some bloat), as far as I can see it only gets used for bad websties that don't use utf8, in which case I think it get delegated to a separate plugin.

Playing with the example and extending it a bit, I kind of see how it works.  A few features I haven't found yet that should exist in the package (I haven't probed to deep yet so maybe I just haven't found them yet... it needs documentation):

1. Selector compaction.  At the moment Matcher.match() returns a list of the selectors... but css doesn't quite work like that, it squishes those selectors down into one using rules to determine how the attributes combine/overwrite each other.
2. Adding/removing elements at run-time.  This should fire of update event callbacks to nodes whose selectors have changed due to selectors, with re-squashed via 1.
3. Editing the CSS at runtime.  This would also fire off events like 2. does.

Not sure if we really need 2. and 3. (perhaps a nice feature for later) but definitely 1, and it shouldn't take too much effort to implement if it doesn't exist already.

What do you think?

**Comment by OceanWolf** [comment]
As an example of point 2. think javascript and for point 3. think firebug

**Comment by JamesRamm** [comment]
Just to throw another angle at this, I think that some fairly major-ish changes are required to MPL before considering the CSS aspects.

Chief among these is separating style handling to its' own reusable object, that can be applied to any of the base `artist`s, probably based on `GraphicsContext`.
This will already provide make graphs far more dynamic in terms of their presentation, at least 'programmatically'. 
After that I'd consider your CSS & figure visitor pattern to apply the CSS. 

to sum up:

1. Detach 'style' from 'artists' to make the figure appearance dynamic
2. Develop a CSS-style syntax and application to the figure to allow figure appearance to be easily scripted.

**Comment by OceanWolf** [comment]
With 1.  I think we already have that with traitlets,
With 2. As I see it we already have the syntax sorted through the packages here.  We just loop over all the elements on draw and query the library for the style to use.

But even if we don't have 1. yet, we still have the update property on the artists to use...By having well thought out modular classes the various steps become independent of each-other, just a thin-bit of glue to lightly connect the parts together.

I think the aim for now lies in getting this css library up to scratch so that we can just plug it into MPL and go.

**Comment by JamesRamm** [comment]
Ok...I guess I havent looked in on MPL for long enough to know about traitlets!

Yes for the CSS...I guess you will need some kind of parser on top of any CSS library you use in order to use MPL-specific selectors etc.. 
You might want to have a look at CartoCSS which is I guess a similar kind of thing that MapBox developed for rendering maps:
https://github.com/mapbox/carto

**Comment by OceanWolf** [comment]
not really a parser, I imagine just handing the css library a list or dictionary or something, and let it get on with it.  Have you seen the library I referenced above, we even have the new owner of said library in this discussion... take a look at
+ https://github.com/SimonSapin/tinycss2 (a generic css parser)
+ https://github.com/SimonSapin/cssselect2 (a generic library for selecting the parsed css that we apply to elements), as you pointed out in the MEP, we can easily build a DOM tree from MPL artists... they have a z-index, and we have elements inside other elements, 

As I say above, they need a bit of tinkering, but they loo mostly there.

**Comment by reedstrm** [comment]
As to the XCSS parsing, both tinycss2 and cssselect2 seem to permissive parsers: they do not enforce most CSS keywords, just the over all 'shape' as it where. When you compile the selector it gets a bit more specific. That's why I just added an extension capability to my fork :-)

As to the use cases against a live tree, the overall design seems to be targeted more at a 'static doc/tree' situation, in particular, there  is a wrapper module that provides a two-way wrapped version of the XML tree. If/when the tree is modified, it needs to be rewrapped. For my use case  (processing huge piles of HTML formatted as a book) that's fine. Not sure about matplotlib.

**Comment by OceanWolf** [comment]
Oh, I thought I saw a list somewhere where one could specify such things.  Anyway, thinking about it more, css works permissively anyway, if a tag, or property, or something does not get recognised, it gets ignored, thus allowing for BC.  So at the most, this checking should just warn, allowing the user to detect a typo, or realise that they use an older-version of the software that uses that CSS (or a newer one that has deprecated those things).  Syntax violations should of course error (as I guess they do).

As for the use cases against a live tree, I take the view that a library should have the generic capacity to do anything, even if not used by the programme the writer wrote the programme for.  After all, the rational behind making something a library comes from allowing share-ability to other programmes who may have other unexpected use-cases.

**Comment by Carreau** [comment]
Long time ago [I did this](https://matthiasbussonnier.com/posts/09-Matplotlib-And-IPython-Config/) which allowed you to do :

```
matplotlib.config.XAxis.Text.t_color='r'
matplotlib.config.YAxis.Text.t_color='b'
```

And would do what you expect.
